### PR TITLE
Drop use of setup.py and setup.cfg

### PR DESCRIPTION
**Pull request by mattclay**
##### SUMMARY

Replace `setup.py` and `setup.cfg` with `pyproject.toml`.

##### ISSUE TYPE

Feature Pull Request

##### ADDITIONAL INFORMATION

Alternative to: https://github.com/ansible/ansible/pull/81440

### CODE REVIEW

**Comment by mattclay** [comment]
The Ubuntu 22.04 failure is apparently due to the use of `pip` before version 22.3, with a venv created with the `--system-site-packages` option. It was fixed with this change: https://github.com/pypa/pip/issues/6264

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,110 @@
...
+"Documentation" = "https://docs.ansible.com/ansible-core/"
+"Mailing lists" = "https://docs.ansible.com/ansible/latest/community/communication.html#mailing-list-information"
+"Source Code" = "https://github.com/ansible/ansible"
+
+[tool.setuptools.dynamic]
+version = {attr = "ansible.release.__version__"}
+dependencies = {file = "requirements.txt"}
+
+[tool.setuptools]
+zip-safe = false
```
Comment (gotmax23) [inline pyproject.toml]: Is `zip-safe` actually needed? We're not building eggs.

Comment (mattclay) [reply pyproject.toml]: I didn't check to see if it was necessary. I just copied it over from the old `setup.cfg`:

https://github.com/ansible/ansible/blob/f10d11bcdc54c9b7edc0111eb38c59a88e396d0a/setup.cfg#L38

Comment (webknjaz) [reply pyproject.toml]: @mattclay

Originally, @AlanCoding added this in https://github.com/ansible/ansible/pull/32194 claiming that it breaks referencing `__file__`. There are still such references in the code base, but I don't know if that statement remains true, or if it'd generate similar tracebacks...

Comment (AlanCoding) [reply pyproject.toml]: It wasn't generating tracebacks if built in the normal way, it just produced warnings.

Anyway, a lot of time has passed and that flag is now considered obsolete since egg and easy_install were deprecated. I expect it to make no difference right now.

Comment (mattclay) [reply pyproject.toml]: I've omitted `zip_safe = False` in the latest rebase.

**Review by nitzmahone — COMMENTED** [review]
Support for a couple of the dynamic attrs in `pyproject.toml` have been marked `BETA` since their inception. TMK they've been very stable, and most folks installing core will just use our wheel anyway, but if we cap setuptools at release-time and (possibly later) ensure that CI is testing with the latest setuptools before we bump it, that should ensure that everything continues to work regardless of what setuptools does with that in the future. Worst case, we can use an in-tree build shim around setuptools if there's an incompatibilty we have to deal with.

**Comment by mattclay** [comment]
https://github.com/ansible/ansible/pull/83713 will take care of the problem of setuptools potentially breaking features we're relying on in pyproject.toml, such as the dynamic dependencies feature, which is still in beta.

File: pyproject.toml [hunk 2]
```diff
@@ -1,3 +1,109 @@
...
+"CI: Azure Pipelines" = "https://dev.azure.com/ansible/ansible/"
+"Documentation" = "https://docs.ansible.com/ansible-core/"
+"Code of Conduct" = "https://docs.ansible.com/ansible/latest/community/code_of_conduct.html"
+
+[tool.setuptools.dynamic]
+version = {attr = "ansible.release.__version__"}
+dependencies = {file = "requirements.txt"}
+
+[tool.setuptools]
+include-package-data = false
```
Comment (gotmax23) [inline pyproject.toml:46]: This ensures that all license files are included in the wheel

```suggestion
include-package-data = false
license-files = [
    "COPYING",
    "licenses/*.txt",
]
```

**Comment by konstruktoid** [comment]
I assume the build process changes a bit, will the docs be updated in this PR as well?

**Comment by webknjaz** [comment]
@konstruktoid no, it doesn't. Per PEP 517, any build front-end will invoke hooks from `setuptools.build_meta`, just like before. If `setup.py` / `setup.cfg` exist, they are taken into account by that backend, but that's transparent and doesn't change anything really.

File: test/sanity/code-smell/package-data.py [hunk 3]
```diff
@@ -36,13 +37,13 @@ def collect_sdist_files(complete_file_list: list[str]) -> list[str]:
...
         'dependency_links.txt',
         'entry_points.txt',
-        'not-zip-safe',
         'requires.txt',
         'top_level.txt',
     )
 
     sdist_files.append('PKG-INFO')
-    sdist_files.extend(f'lib/ansible_core.egg-info/{name}' for name in egg_info)
+    sdist_files.append('setup.cfg')
```
Comment (webknjaz) [inline test/sanity/code-smell/package-data.py:46]: Why is this added? Should it be dropped?
```suggestion
```

Comment (mattclay) [reply test/sanity/code-smell/package-data.py:46]: It's generated during the build and included in the sdist, which is why it's added to the list of files to expect.

**Review by sivel — APPROVED** [review]
> Support for a couple of the dynamic attrs in `pyproject.toml` have been marked `BETA` since their inception

I had this on my mind too, so I went to check the docs, and I think we might both be confused about their support in `setup.cfg` as opposed to `pyproject.toml`.  `setup.cfg` has a number of things marked as beta, but I read the whole docs page about `pyproject.toml` and everything is supported, and I see nothing marked as beta functionality.  

With that being said, I have no concerns with this PR.  LGTM

**Comment by mattclay** [comment]
@sivel There are at least two features to pay attention to:

* https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#setuptools-specific-configuration -- specifically `license-files` (marked `Provisional`), which we're not using in this PR, but it was suggested.
* https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata -- specifically `dependencies` which is marked `Beta`.

**Comment by sivel** [comment]
Hm, that is super weird. I explicitly searched the page for beta, and had no matches, and I read it carefully.  Maybe I'm just dumb.  I see them now.  I'm still ok with the `dependencies` use.

**Comment by mattclay** [comment]
We do need to keep the project requirements in sync with the ones embedded in ansible-test. Currently that's easy to do because we maintain two identical copies of the same file. If we inline the requirements we'll need to update our tests so they're able to compare the inline requirements against the ones ansible-test uses.

File: test/sanity/code-smell/package-data.py (hunk 3 above)
Comment (webknjaz) [reply test/sanity/code-smell/package-data.py:46]: Oh, is that for some ancient envs?

**Comment by mattclay** [comment]
@sivel Any thoughts on whether or not we should incorporate the suggested `license-files` change?

**Comment by webknjaz** [comment]
This setting originated from `wheel` many years ago, and it was customary to have it in `setup.cfg`. Having been moved to `pyproject.toml`, shouldn't change anything.

P.S. I know there's work happening to finalize the PEP for standardizing license inclusion, led by redhatters these days. But it'll be some time until that gets implemented.

File: test/sanity/code-smell/package-data.py (hunk 3 above)
Comment (mattclay) [reply test/sanity/code-smell/package-data.py:46]: It appears to be the default behavior in setuptools. See https://github.com/pypa/setuptools/blob/46fe69938e2b88fcad78d2dfbcfc701a4cfa6b3f/setuptools/command/sdist.py#L174 and https://github.com/pypa/setuptools/blob/46fe69938e2b88fcad78d2dfbcfc701a4cfa6b3f/setuptools/command/egg_info.py#L207
